### PR TITLE
Fix #1134 and #906: Support variable length ISO8601 dates

### LINKED ISSUE #906: ISO8601 datetime format has a variable length, we expect constant length
Throughout our code we use iso8601 datetieme format, made by `maya`. Maya returns it with microseconds, unless number of microseconds is 0. So, the result has a variable length:
```
In [2]: t = maya.now()                                                        

In [3]: t.iso8601()                                                           
Out[3]: '2019-04-03T22:15:05.400470Z'

In [4]: maya.MayaDT(t.epoch).iso8601()                                        
Out[4]: '2019-04-03T22:15:05Z'
```

If we pass the datetime in "0 microseconds" format to our API, it will result in Ursula failing to parse.
Example: replace `(maya.now() + datetime.timedelta(days=3)).iso8601()` with `maya.MayaDT((maya.now() + datetime.timedelta(days=3)).epoch).iso8601()` for expiration in `tests/characters/test_character_control.py`

The result:
```
  File "/home/michwill/Projects/nucypher-kms/nucypher/network/server.py", line 221, in consider_arrangement
    arrangement = Arrangement.from_bytes(request.data)
  File "/home/michwill/Projects/nucypher-kms/nucypher/policy/models.py", line 89, in from_bytes
    alice_pubkey_sig, arrangement_id, expiration_bytes = cls.splitter(arrangement_as_bytes)
  File "/home/michwill/.local/share/virtualenvs/nucypher-kms-5SSQ9ZHE/lib/python3.6/site-packages/bytestring_splitter/__init__.py", line 57, in __call__
    self.message_types, len(self), len(splittable)))
bytestring_splitter.BytestringSplittingError: Wrong number of bytes to constitute message types [(None, <class 'umbral.keys.UmbralPublicKey'>, 33, {}), (None, <class 'bytes'>, 32, {}), (None, <class 'bytes'>, 27, {})] - 
                    need 92, got 85 
 Did you mean to return the remainder?
```

### LINKED ISSUE #1134: Grant fails with logger KeyError

### PR DESCRIPTION
**Pull request by cygnusv**
It seems that the old issue of variable length ISO8601 dates (#906) was behind #1134.

For the record, this is the grant command that previously failed as in #1134, but now works after this fix:
```
nucypher alice grant --teacher-uri 127.0.0.1:11500 --bob-verifying-key 0277b5f24dbd68d4cdbbb9f54d6b892b52a48513695be4f703c2516a1f94e1dc9e --bob-encrypting-key 0277b5f24dbd68d4cdbbb9f54d6b892b52a48513695be4f703c2516a1f94e1dc9e --label asdf --expiration 2019-12-30T10:07:50Z --m 1 --n 1 --dev
```
Tested with a lonely dev ursula for the moment. The output seems to show everything is OK:
```

    / \  | (_) ___ ___
   / _ \ | | |/ __/ _ \
  / ___ \| | | (_|  __/
 /_/   \_|_|_|\___\___|

 the Authority.

WARNING: Running in Federated mode
treasure_map ...... MPoQ6eWorfP/lCRyODElVveCVEA/3Ed0q/RmJ2QCNTlwhqLpkSUn7EzpQ8sRqkKyxhkhduAwEYngN0akSNVpJjmJB6ppU1CztimP3PG4BHYCxO8X9cPiizAGiXc+o6xiAAABHANgefwHV3Aj5WrUoIMPJiHVGEvoUBOR5ckp2757yHAemAI37rO8Qy+lg6p32HvmQO4fb4s0jBFO5151y3yMZZTHojHj5KODWg9Rdm0DZZvKZ2dzD85InvQ4+z04NRJ3lXaMA73EUXTGq+WfIH4ZPvyeyeFgUQ8tfQWcQaCRO12+2aPt48SebLTxuOQfbBNFgXLa8WUHnSwbC2mt1mFa37fPHDQ0C4CFW9F9ywyprl5pkzfFIY9OVWdmvMkQrrwopQbjO5/NMdlYHQ/7YEPKXDSAarIbh4rQoGL54j0Bv0MlOGpCW+TXuxs7GGzumNIf+c44A8k3djoGFj0Z+XPWb5ygEH6pyE7kwb0jKcjIde+NlC2INpToxaJ+pdzw
policy_encrypting_key ...... 025c5f763863da6018447498af61b443f23c63dd7699502f0bcfd843dbda1274dd
alice_verifying_key ...... 03bdc45174c6abe59f207e193efc9ec9e160510f2d7d059c41a0913b5dbed9a3ed
request_id ...... 1564487069
duration ...... 0:00:01
```
-----
* Fixes #906.
* Fixes #1134.

### CODE REVIEW

File: nucypher/characters/control/serializers.py [hunk 1]
```diff
@@ -86,7 +86,7 @@ def load_create_policy_input(request: dict):
                             label=request['label'].encode(),
                             m=request['m'],
                             n=request['n'],
-                            expiration=request['expiration'])
+                            expiration=maya.MayaDT.from_iso8601(iso8601_string=request['expiration']))
```
Comment (tuxxy) [inline nucypher/characters/control/serializers.py]: :+1:

File: nucypher/policy/models.py [hunk 2]
```diff
@@ -57,8 +56,9 @@ class Arrangement:
     federated = True
     ID_LENGTH = 32
 
-    splitter = key_splitter + BytestringSplitter((bytes, ID_LENGTH),
-                                                 (bytes, 27))
+    splitter = BytestringSplitter((UmbralPublicKey, PUBLIC_KEY_LENGTH),  # alice.stamp
+                                  (bytes, ID_LENGTH),  # arrangement_ID
+                                  (bytes, VariableLengthBytestring))  # expiration
```
Comment (tuxxy) [inline nucypher/policy/models.py]: :+1:

File: nucypher/policy/models.py [hunk 3]
```diff
@@ -75,10 +75,14 @@ def __init__(self,
 
         Other params are hopefully self-evident.
         """
-        self.id = arrangement_id or secure_random(self.ID_LENGTH)
+        if arrangement_id:
+            if len(arrangement_id) != self.ID_LENGTH:
+                raise ValueError(f"Arrangement ID must be of length {self.ID_LENGTH}.")
```
Comment (tuxxy) [inline nucypher/policy/models.py]: Good validation. :+1:

File: nucypher/characters/control/emitters.py [hunk 4]
```diff
@@ -46,6 +46,8 @@ def _emit(self,
               message: str = None,
               color: str = None,
               bold: bool = False,
+              request_id: int = -1,  # TODO: What to do when request_id and duration are received?
```
Comment (tuxxy) [inline nucypher/characters/control/emitters.py]: I don't really understand the purpose of this, can you elaborate?

Comment (cygnusv) [reply nucypher/characters/control/emitters.py]: I can't! As the commit says, it's only a workaround. If you see the signatures of the `_emit` methods in other emitters, there are a `request_id` and `duration` arguments that this method doesn't have and it fails when it receives them. But I don't really don't know what I should with them, hence the comments.

Comment (tuxxy) [reply nucypher/characters/control/emitters.py]: haha gotcha

**Comment by KPrasch** [comment]
Needs rebase.  Even if we dont settle on this format, it does fix the bug.

**Comment by cygnusv** [comment]
Sure, I'll rebase it now. The emitters part seem to have changed a lot too.

**Pull request merged**